aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -17,21 +17,74 @@
 #include "aesd-circular-buffer.h"
 
 /**
- * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
- * @param char_offset the position to search for in the buffer list, describing the zero referenced
- *      character index if all buffer strings were concatenated end to end
- * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
- *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
- *      in aesd_buffer.
+ * @param buffer the buffer to search for corresponding offset. Any necessary locking must be performed by the caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero-referenced
+ *      character index if all buffer strings were concatenated end to end.
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte offset of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset. This value is only set when a matching char_offset is found
+ *      in the aesd_buffer.
  * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
  * NULL if this position is not available in the buffer (not enough data is written).
  */
 struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
             size_t char_offset, size_t *entry_offset_byte_rtn )
 {
-    /**
-    * TODO: implement per description
-    */
+    // Check for invalid input, return NULL if input is invalid
+    if (buffer == NULL)
+    {
+        return NULL;
+    }
+    
+    if (entry_offset_byte_rtn == NULL)
+    {
+        return NULL;
+    }
+
+    size_t total_size = 0;  // Initialize total size of characters
+    uint8_t index = buffer->out_offs;  // Start from the oldest entry
+    uint8_t count;
+
+    // find how many entries are in the buffer
+    if (buffer->full)
+    {
+        count = AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    }
+    else
+    {
+        count = buffer->in_offs - buffer->out_offs;
```

Comment:
This assumes `in_offs` is > `out_offs` when the buffer is not full, which won't always be the case.